### PR TITLE
feat: enable rust binding for `TwoAdicFriPcs::Verify()`

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This PR enables rust binding for `TwoAdicFriPcs::Verify()`.

### CODE REVIEW

File: tachyon/crypto/commitments/fri/two_adic_fri.h [hunk 1]
```diff
@@ -231,66 +221,77 @@ class TwoAdicFRI {
...
                   const ExtF& p_at_z = ps_at_z[j];
-                  ExtF quotient =
-                      unwrap((ExtF(opened_values[batch_idx][j]) - p_at_z) /
-                             (ExtF(x) - z));
-                  std::tuple<ExtF, ExtF>& reduced_opening =
-                      reduced_openings[log_num_rows];
-                  std::get<1>(reduced_opening) +=
-                      std::get<0>(reduced_opening) * quotient;
-                  std::get<0>(reduced_opening) *= alpha;
+                  ExtF quotient = (ExtF(cur_values_in[j]) - p_at_z) * denom;
```
Comment (GideokKim) [inline tachyon/crypto/commitments/fri/two_adic_fri.h:292]:
```suggestion
                  ExtF quotient = (ExtF(cur_values_in[j]) - ps_at_z[j]) * denom;
```

File: tachyon/c/zk/air/sp1/baby_bear_poseidon2_two_adic_fri.h [hunk 2]
```diff
@@ -111,6 +113,27 @@ TACHYON_C_EXPORT void tachyon_sp1_baby_bear_poseidon2_two_adic_fri_open(
     tachyon_sp1_baby_bear_poseidon2_opened_values** opened_values,
     tachyon_sp1_baby_bear_poseidon2_fri_proof** proof);
 
+/**
+ * @brief Verifies an opening proof with commitments,
```
Comment (TomTaehoonKim) [inline tachyon/c/zk/air/sp1/baby_bear_poseidon2_two_adic_fri.h:117]:
```suggestion
 * @brief Verifies an opening proof with commitments.
```

File: tachyon/c/zk/air/sp1/baby_bear_poseidon2_opened_values.cc [hunk 3]
```diff
@@ -12,8 +13,8 @@ using OpenedValues =
     std::vector<std::vector<std::vector<std::vector<math::BabyBear4>>>>;
 
 tachyon_sp1_baby_bear_poseidon2_opened_values*
-tachyon_sp1_baby_bear_poseidon2_opened_values_create() {
-  return c::base::c_cast(new OpenedValues());
+tachyon_sp1_baby_bear_poseidon2_opened_values_create(size_t rounds) {
+  return c::base::c_cast(new OpenedValues(rounds));
```
Comment (TomTaehoonKim) [inline tachyon/c/zk/air/sp1/baby_bear_poseidon2_opened_values.cc:17]: I don't get why this change is in this commit.

File: tachyon/c/zk/air/sp1/baby_bear_poseidon2_opened_values.cc [hunk 4]
```diff
@@ -40,3 +41,32 @@ void tachyon_sp1_baby_bear_poseidon2_opened_values_serialize(
...
+void tachyon_sp1_baby_bear_poseidon2_opened_values_allocate_inner(
+    tachyon_sp1_baby_bear_poseidon2_opened_values* opened_values, size_t round,
+    size_t r, size_t cols, size_t size) {
+  std::vector<std::vector<std::vector<math::BabyBear4>>>& mat =
+      c::base::native_cast(*opened_values)[round];
+  std::vector<std::vector<math::BabyBear4>>& row = mat[r];
+  for (size_t c = 0; c < cols; ++c) {
+    row[c].resize(size);
+  }
+}
```
Comment (TomTaehoonKim) [inline tachyon/c/zk/air/sp1/baby_bear_poseidon2_opened_values.cc:65]: Shouldn't this be in the next commit?

File: tachyon/c/zk/air/sp1/baby_bear_poseidon2_opened_values.cc (hunk 3 above)
Comment (chokobole) [reply tachyon/c/zk/air/sp1/baby_bear_poseidon2_opened_values.cc:17]: All `xxx_allocate()` C APIs accept a `size_t` argument for round, which indicates the index of the round. This implies that these vectors are expected to be resized according to rounds beforehand. Similarly, `tachyon_sp1_baby_bear_poseidon2_opened_values_allocate_outer()` and `tachyon_sp1_baby_bear_poseidon2_opened_values_allocate_inner()` require that the vectors be resized during construction.

File: tachyon/c/zk/air/sp1/baby_bear_poseidon2_opened_values.cc (hunk 4 above)
Comment (chokobole) [reply tachyon/c/zk/air/sp1/baby_bear_poseidon2_opened_values.cc:65]: That's also possible, however by separating out the Rust binding changes from the C API changes, I can make it easier for reviewers to focus specifically on the C API modifications.

**Review by GideokKim — APPROVED** [review]
LGTM

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Review by dongchangYoo — APPROVED** [review]
LGTM

File: tachyon/crypto/commitments/fri/two_adic_fri.h [hunk 5]
```diff
@@ -265,19 +265,19 @@ class TwoAdicFRI {
...
               std::tuple<ExtF, ExtF>& reduced_opening = it.first->second;
               for (size_t i = 0; i < cur_points.size(); ++i) {
                 const ExtF& z = cur_points[i];
+                ExtF denom = unwrap((x - z).Inverse());
                 const std::vector<ExtF>& ps_at_z = cur_values[i];
                 CHECK_EQ(ps_at_z.size(), cur_values_in.size());
                 for (size_t j = 0; j < ps_at_z.size(); ++j) {
-                  ExtF quotient = unwrap((ExtF(cur_values_in[j]) - ps_at_z[j]) /
-                                         (ExtF(x) - z));
+                  ExtF quotient = (ExtF(cur_values_in[j]) - ps_at_z[j]) * denom;
```
Comment (ashjeong) [inline tachyon/crypto/commitments/fri/two_adic_fri.h:291]: What about also minimizing conversions from f to extf by initializing cur_values_in to extfs

File: tachyon/c/zk/air/sp1/baby_bear_poseidon2_domains.h [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+#include "tachyon/c/math/finite_fields/baby_bear/baby_bear.h"
+
+struct tachyon_sp1_baby_bear_poseidon2_domains {};
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Creates a new domains.
```
Comment (ashjeong) [inline tachyon/c/zk/air/sp1/baby_bear_poseidon2_domains.h:23]:
```suggestion
 * @brief Creates a new set of domains.

File: tachyon/c/zk/air/sp1/baby_bear_poseidon2_domains.h [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+ *
+ * @param rounds The number of rounds.
+ * @return A pointer to the newly created domains.
+ */
+TACHYON_C_EXPORT
+tachyon_sp1_baby_bear_poseidon2_domains*
+tachyon_sp1_baby_bear_poseidon2_domains_create(size_t rounds);
+
+/**
+ * @brief Destroys a domains, freeing its resources.
```
Comment (ashjeong) [inline tachyon/c/zk/air/sp1/baby_bear_poseidon2_domains.h:33]:
```suggestion
 * @brief Destroys a set of domains, freeing its resources.

File: tachyon/c/zk/air/sp1/baby_bear_poseidon2_domains.h [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+ * @param domains A pointer to the domains.
+ * @param round The round.
+ * @param size The size of the domains.
+ */
+TACHYON_C_EXPORT void tachyon_sp1_baby_bear_poseidon2_domains_allocate(
+    tachyon_sp1_baby_bear_poseidon2_domains* domains, size_t round,
+    size_t size);
+
+/**
+ * @brief Sets domain.
```
Comment (ashjeong) [inline tachyon/c/zk/air/sp1/baby_bear_poseidon2_domains.h:52]:
```suggestion
 * @brief Sets up a domain.

File: tachyon/c/zk/air/sp1/baby_bear_poseidon2_commitment_vec.h [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+ * @return A pointer to the newly created commitment vector.
+ */
+TACHYON_C_EXPORT
+tachyon_sp1_baby_bear_poseidon2_commitment_vec*
+tachyon_sp1_baby_bear_poseidon2_commitment_vec_create(size_t rounds);
+
+/**
+ * @brief Destroys a commitment vector, freeing its resources.
+ *
+ * @param commitment vector A pointer to the commitment vector to destroy.
```
Comment (ashjeong) [inline tachyon/c/zk/air/sp1/baby_bear_poseidon2_commitment_vec.h:35]:
```suggestion
 * @param commitment _vec A pointer to the commitment vector to destroy.

File: tachyon/c/zk/air/sp1/baby_bear_poseidon2_commitment_vec.h [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+ */
+TACHYON_C_EXPORT void tachyon_sp1_baby_bear_poseidon2_commitment_vec_destroy(
+    tachyon_sp1_baby_bear_poseidon2_commitment_vec* commitment_vec);
+
+/**
+ * @brief Sets commitment.
+ *
+ * @param commitment vector A pointer to the commitment vector.
```
Comment (ashjeong) [inline tachyon/c/zk/air/sp1/baby_bear_poseidon2_commitment_vec.h:43]:
```suggestion
 * @param commitment_vec A pointer to the commitment vector.

**Review by batzor — APPROVED** [review]
LGTM

**Review by ashjeong — COMMENTED** [review]
Can you squash together  https://github.com/kroma-network/tachyon/pull/532/commits/1c9bfeeb7fe9ed3f47f8b78d8e527fc706dd5774 and https://github.com/kroma-network/tachyon/pull/532/commits/1e02ede6e0307a4ca67986711e2b1a1da57e92c7?
